src/workerd/api/streams/internal.c++

```diff
@@ -96,23 +94,23 @@ private:
   kj::Vector<kj::Array<kj::byte>> parts;
   uint64_t runningTotal = 0;
 
-  kj::Promise<PartList> loop() {
-    auto bytes = kj::heapArray<kj::byte>(4096);
+  kj::Promise<PartList> readParts() {
+    kj::byte buffer[4096];
+
+    while (true) {
+      auto amount = co_await input.tryRead(buffer, 1, kj::size(buffer));
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
       if (amount == 0) {
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+        co_return KJ_MAP(p, parts) { return p.asPtr(); };
       }
 
       runningTotal += amount;
       if (runningTotal >= limit) {
-        return JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
+        throw JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
       }
-      parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      return loop();
-    });
+
+      parts.add(kj::heapArray<kj::byte>(buffer, amount));
```

Comment:
Hmm, it's too bad this does an extra copy of the data -- although the old code had a different problem, it might have allocated a 4k buffer, read 1 bytes into it, and just kept the rest as wasted space.

Since we're ultimately returning heap buffers, perhaps we should actually allocate our buffer on the heap upfront, and then do `tryRead()` with `minBytes == maxBytes`. This way, the read won't return until either EOF or the buffer is full, so we get the best of both worlds I think.

Reply:
yeah, I was thinking the same thing. I wonder how much data is read in practice? One packet? I'll try this out with minBytes == maxBytes tomorrow.